main/src/library/Result.flix

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright 2016 Liam Palmer
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Result {
+
+    ///
+    /// Returns `true` iff `r` is `Ok(v)`.
+    ///
+    def isOk[t, e](r: Result[t, e]): Bool = match r with {
+        case Ok(v) => true
+        case Err(w) => false
+    }
+
+    ///
+    /// Returns `true` iff `r` is `Err(w)`.
+    ///
+    def isErr[t, e](r: Result[t, e]): Bool = match r with {
+        case Ok(v) => false
+        case Err(w) => true
+    }
+
+    ///
+    /// Returns `v` if `r` is `Ok(v)`. Undefined semantics if `r` is `Err(w)`.
+    ///
+    def get[t, e](r: Result[t, e]): t = match r with {
+        case Ok(v) => v
+    }
+
+    ///
+    /// Returns `v` if `r` is `Ok(v)`. Otherwise returns `d`.
+    ///
+    def getWithDefault[t, e](r: Result[t, e], d: t): t = match r with {
+        case Ok(v) => v
+        case Err(w) => d
+    }
+
+    ///
+    /// Returns `r1` if it is `Ok(v)`. Otherwise returns `r2`.
+    ///
+    def withDefault[t, e1, e2](r1: Result[t, e1], r2: Result[t, e2]): Result[t, e2] = match r1 with {
+        case Ok(v) => r1
+        case Err(w) => r2
+    }
+
+    ///
+    /// Returns `true` if `r` is `Ok(v)` and the predicate `f(v)` evaluates to `true`. Otherwise returns `false`.
+    ///
+    def exists[t, e](f: t -> Bool, r: Result[t, e]): Bool = getWithDefault(map(f, r), false)
+
+    ///
+    /// Returns `true` if `r` is `Ok(v)` and the predicate `f(v)` evaluates to `true` or if `r` is `Err(w)`.
+    /// Otherwise returns `false`.
+    ///
+    def forall[t, e](f: t -> Bool, r: Result[t, e]): Bool = getWithDefault(map(f, r), true)
+
+    ///
+    /// Returns `r` if `r` is `Ok(v)` and the predicate `f(v)` evaluates to true. Otherwise returns `Err(d)`.
+    ///
+    def filter[t, e](f: t -> Bool, r: Result[t, e], d: e): Result[t, e] = match r with {
+        case Ok(v) => if (f(v)) r else Err(d)
+        case Err(w) => Err(d)
+    }
+
+    ///
+    /// Returns `Ok(f(v))` if `r` is `Ok(v)`. Otherwise returns `r`.
+    ///
+    def map[t1, t2, e](f: t1 -> t2, r: Result[t1, e]): Result[t2, e] = match r with {
+        case Ok(v) => Ok(f(v))
+        case Err(w) => r
+    }
+
+    ///
+    /// Returns `Ok(f(v1, v2))` if `r1` is `Ok(v1)` and `r2` is `Ok(v2)`.
+    /// Otherwise returns the first of `r1` and `r2` that is `Err(w)`.
+    ///
+    def map2[t1, t2, t3, e](f: (t1, t2) -> t3, r1: Result[t1, e], r2: Result[t2, e]): Result[t3, e] = match (r1, r2) with {
```

Comment:
Delete. `map2` doesn't make sense when we have two errors; which one would we get? The first? The second? I think it is best to remove it.